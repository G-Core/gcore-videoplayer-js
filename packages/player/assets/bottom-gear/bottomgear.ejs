<button type="button" class="button-gear media-control-button gplayer-lite-btn gcore-skin-button-color gear-icon" data-gear-button="-1" id="gear-button">
    <%= icon %>
</button>
<div class="gear-wrapper gcore-skin-bg-color">
    <ul class="gear-options-list" id="gear-options">
        <% for (const item of items) { %>
            <li data-<%= item %>></li>
        <% } %>
    </ul>
</div>