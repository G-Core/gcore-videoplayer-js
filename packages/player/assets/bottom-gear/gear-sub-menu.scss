*, :focus, :visited {
  outline: none !important;
}

.gear-wrapper {
  .go-back {
    font-weight: 600;
    font-size: 14px;
    line-height: 20px;
    width: 100%;
    text-align: left;
    padding: 12px;

    .arrow-left-icon {
      float: left;
      padding-top: 2px;
      padding-right: 2px;

      svg {
        height: 16px;
      }
    }
  }

  ul.gear-sub-menu {
    width: 100%;
    list-style-type: none;
    min-width: 60px;
    border-top: 2px solid rgb(36 36 36);
    overflow-y: auto;

    li {
      font-size: 12px;
      text-align: left;

      &[data-title] {
        background-color: #c3c2c2;
        padding: 5px;
      }

      a {
        display: block;
        text-decoration: none;
        height: 32px;
        padding: 5px 10px;
        line-height: 22px;
        color: #fffffe;

        &:hover {
          color: white;
          background-color: rgb(0 0 0 / 40%);

          a {
            color: white;
            text-decoration: none;
          }
        }

        .check-icon {
          width: 30px;
          height: 20px;
          float: left;
          display: block;

          svg {
            display: none;
          }
        }
      }

      &.current a {
        .check-icon {
          svg {
            display: inline;
          }
        }
      }
    }
  }
}
