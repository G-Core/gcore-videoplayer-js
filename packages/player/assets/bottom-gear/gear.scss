.media-control-skin-1 {
  .media-control-item.media-control-gear {
    order: 99;

    .gear-wrapper {
      position: absolute;
      right: 16px;
      bottom: 52px;

      width: 250px;
      min-height: 48px;
      z-index: 9999;
      border-radius: 4px;
      box-shadow: 0 0 1px 1px var(--gplayer-mc-popup-border-color);
    }

    .gear-options-list {
      padding: 8px 0;
    }

    .gear-option {
      margin: 0;
      text-align: left;
      line-height: 22px;
      padding: 5px 14px;
      width: 250px;
      font-size: 12px;
      display: flex;
      align-items: center;
      justify-content: flex-start;
      gap: 8px;

      &_icon {
        flex: 24px 0 0;
        height: 24px;

        &.hidden {
          visibility: hidden;
          display: inline-block;
        }
      }

      &_label {
        flex: 0 1 100%;
      }

      &_arrow-right-icon {
        flex: 0 0 14px;
        height: 24px;
      }

      &_value {
        flex: 1 0 auto;
      }
    }
  }
}