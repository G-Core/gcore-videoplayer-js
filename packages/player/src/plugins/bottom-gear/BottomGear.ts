import { UICorePlugin, template, Events as ClapprEvents, $ } from '@clappr/core'
import { trace } from '@gcorevideo/utils'
import assert from 'assert'

import { CLAPPR_VERSION } from '../../build.js'

import pluginHtml from '../../../assets/bottom-gear/bottomgear.ejs'
import '../../../assets/bottom-gear/gear.scss'
import '../../../assets/bottom-gear/gear-sub-menu.scss'
import gearIcon from '../../../assets/icons/new/gear.svg'
import gearHdIcon from '../../../assets/icons/new/gear-hd.svg'
import { ZeptoResult } from '../../types.js'

const VERSION = '2.19.12'

const T = 'plugins.bottom_gear'

/**
 * Events triggered by the plugin
 * @beta
 */
export enum GearEvents {
  /**
   * Subscribe to this event to accurately attach an item to the gear menu
   */
  RENDERED = 'rendered',
}

// TODO disabled if no items added

/**
 * `PLUGIN` that adds a button to extend the media controls UI with extra options.
 * @beta
 * @remarks
 * The plugin renders small gear icon to the right of the media controls.
 * It provides a base for attaching custom settings UI in the gear menu
 *
 * Depends on:
 *
 * - {@link MediaControl}
 *
 * @example
 * You can use bottom gear to add custom settings UI to the gear menu.
 *
 * ```ts
 * import { BottomGear } from '@gcorevideo/player/plugins/bottom-gear';
 *
 * class CustomOptionsPlugin extends UICorePlugin {
 *   // ...
 *
 *   override get events() {
 *     return {
 *       'click #my-button': 'doMyAction',
 *     }
 *   }
 *
 *   private doMyAction() {
 *     // ...
 *   }
 *
 *   override render() {
 *     const bottomGear = this.core.getPlugin('bottom_gear');
 *     if (!bottomGear) {
 *       return this;
 *     }
 *     this.$el.html('<button class="custom-option">Custom option</button>');
 *     // Put rendered element into the gear menu
 *     bottomGear.addItem('custom').html(this.$el)
 *     return this;
 *   }
 *
 *   // alternatively, add an option with a submenu
 *   override render() {
 *     this.$el.html(template(templateHtml)({
 *       // ...
 *     })));
 *     return this;
 *   }
 *
 *   private addGearOption() {
 *     this.core.getPlugin('bottom_gear')
 *       .addItem('custom', this.$el)
 *       .html($('<button class="custom-option">Custom option</button>'))
 *   }
 *
 *   override bindEvents() {
 *     this.listenToOnce(this.core, ClapprEvents.CORE_READY, () => {
 *       const bottomGear = this.core.getPlugin('bottom_gear');
 *       assert(bottomGear, 'bottom_gear plugin is required');
 *       // simple case
 *       this.listenTo(bottomGear, GearEvents.RENDERED, this.render);
 *       // or with a submenu
 *       this.listenTo(bottomGear, GearEvents.RENDERED, this.addGearOption);
 *      });
 *   }
 * }
 * ```
 */
export class BottomGear extends UICorePlugin {
  private isHd = false

  /**
   * @internal
   */
  get name() {
    return 'bottom_gear'
  }

  /**
   * @internal
   */
  get supportedVersion() {
    return { min: CLAPPR_VERSION }
  }

  /**
   * @internal
   */
  static get version() {
    return VERSION
  }

  private static readonly template = template(pluginHtml)

  /**
   * @internal
   */
  override get attributes() {
    return {
      class: 'media-control-gear',
    }
  }

  /**
   * @internal
   */
  override get events() {
    return {
      'click #gear-button': 'toggleGearMenu',
    }
  }

  /**
   * @internal
   */
  override bindEvents() {
    this.listenToOnce(this.core, ClapprEvents.CORE_READY, this.onCoreReady)
    this.listenTo(
      this.core,
      ClapprEvents.CORE_ACTIVE_CONTAINER_CHANGED,
      this.onActiveContainerChanged,
    )
  }

  /**
   * Adds a custom option to the gear menu
   * @param name - A unique name of the option
   * @param $subMenu - The submenu to attach to the option
   * @returns The added item placeholder to attach custom markup
   * @remarks
   * When called with $submenu param, a click on the added item will toggle the submenu visibility.
   *
   * When added without submenu, it's responsibility of the caller to handle the click event however needed.
   * @example
   * ```ts
   * class MyPlugin extends UICorePlugin {
   *   override render() {
   *     this.$el.html('<div class="my-awesome-settings">...</div>')
   *     this.core.getPlugin('bottom_gear')
   *       ?.addItem('custom', this.$el)
   *       .html($('<button>Custom settings</button>'))
   *     return this
   *   }
   * }
   * ```
   */
  addItem(name: string, $subMenu?: ZeptoResult): ZeptoResult {
    const $existingItem = this.$el.find(`#gear-options li[data-${name}`)
    if ($existingItem.length) {
      trace(`${T} addItem already exists`, { name })
      return $existingItem
    }
    const $item = $('<li></li>')
      .attr(`data-${name}`, '')
      .appendTo(this.$el.find('#gear-options'))
    if ($subMenu) {
      trace(`${T} addItem adding submenu`, { name })
      $subMenu
        .addClass('gear-sub-menu-wrapper')
        .hide()
        .appendTo(this.$el.find('#gear-options-wrapper'))
      $item.on('click', (e: MouseEvent) => {
        trace(`${T} addItem submenu clicked`, { name })
        e.stopPropagation()
        $subMenu.show()
        this.$el.find('#gear-options').hide()
      })
    }
    return $item
  }

  private onActiveContainerChanged() {
    trace(`${T} onActiveContainerChanged`)
    this.bindContainerEvents()
  }

  private bindContainerEvents() {
    trace(`${T} bindContainerEvents`)
    this.listenTo(
      this.core.activeContainer,
      ClapprEvents.CONTAINER_HIGHDEFINITIONUPDATE,
      this.highDefinitionUpdate,
    )
  }

  private highDefinitionUpdate(isHd: boolean) {
    trace(`${T} highDefinitionUpdate`, { isHd })
    this.isHd = isHd
    this.$el.find('.gear-icon').html(isHd ? gearHdIcon : gearIcon)
  }

  /**
   * @internal
   */
  override render() {
    trace(`${T} render`)
    const mediaControl = this.core.getPlugin('media_control')
    if (!mediaControl) {
      return this // TODO test
    }
    const icon = this.isHd ? gearHdIcon : gearIcon
    this.$el
      .html(BottomGear.template({ icon }))
      .find('#gear-sub-menu-wrapper')
      .hide()

    // TODO make non-clickable when there are no items

    setTimeout(() => {
      this.trigger(GearEvents.RENDERED)
    }, 0)

    return this
  }

  /**
   * Collapses any submenu open back to the gear menu.
   * @remarks
   * Should be called by the UI plugin that added a gear item with a submenu when the latter is closed (e.g., when a "back" button is clicked).
   */
  refresh() {
    this.$el.find('.gear-sub-menu-wrapper').hide()
    this.$el.find('#gear-options').show()
  }

  private toggleGearMenu() {
    this.$el.find('#gear-options-wrapper').toggle()
  }

  private hide() {
    this.$el.find('#gear-options-wrapper').hide()
  }

  private onCoreReady() {
    trace(`${T} onCoreReady`)
    const mediaControl = this.core.getPlugin('media_control')
    assert(mediaControl, 'media_control plugin is required')
    this.listenTo(
      mediaControl,
      ClapprEvents.MEDIACONTROL_RENDERED,
      this.onMediaControlRendered,
    )
    this.listenTo(mediaControl, ClapprEvents.MEDIACONTROL_HIDE, this.hide)
  }

  private onMediaControlRendered() {
    trace(`${T} onMediaControlRendered`)
    const mediaControl = this.core.getPlugin('media_control')
    mediaControl.mount('gear', this.$el)
  }
}
