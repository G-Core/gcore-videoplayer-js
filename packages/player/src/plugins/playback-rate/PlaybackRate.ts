import { Events, UICorePlugin, Playback, template, Core } from '@clappr/core'
import { trace } from '@gcorevideo/utils'
import assert from 'assert'

import { CLAPPR_VERSION } from '../../build.js'
import type { ZeptoResult } from '../../types.js'

import buttonHtml from '../../../assets/playback-rate/button.ejs'
import listHtml from '../../../assets/playback-rate/list.ejs'
import speedIcon from '../../../assets/icons/new/speed.svg'
import arrowRightIcon from '../../../assets/icons/new/arrow-right.svg'
import arrowLeftIcon from '../../../assets/icons/new/arrow-left.svg'
import checkIcon from '../../../assets/icons/new/check.svg'
import { BottomGear } from '../bottom-gear/BottomGear.js'
import { PlaybackEvents } from '../../playback/types.js'
import {
  MediaControl,
  MediaControlEvents,
} from '../media-control/MediaControl.js'

type PlaybackRateOption = {
  value: string
  label: string
}

const DEFAULT_PLAYBACK_RATES = [
  { value: '0.5', label: '0.5x' },
  { value: '0.75', label: '0.75x' },
  { value: '1.0', label: '1x' },
  { value: '1.25', label: '1.25x' },
  { value: '1.5', label: '1.5x' },
  { value: '1.75', label: '1.75x' },
  { value: '2.0', label: '2x' },
]

const DEFAULT_PLAYBACK_RATE = '1.0'

const T = 'plugins.playback_rate'

/**
 * `PLUGIN` that allows changing the playback speed of the video.
 * @beta
 *
 * @remarks
 * Depends on:
 *
 * - {@link MediaControl | media_control}
 *
 * - {@link BottomGear | bottom_gear}
 *
 * It renders a button in the gear menu, which opens a dropdown with the options to change the playback rate.
 * Note that the playback rate change is supported only for VOD or DVR enabled live streams.
 */
export class PlaybackRate extends UICorePlugin {
  private playbackRates: PlaybackRateOption[] = DEFAULT_PLAYBACK_RATES

  // Saved when an ad starts to restore after it finishes
  private prevSelectedRate: string | undefined

  private rendered = false

  private selectedRate: string = DEFAULT_PLAYBACK_RATE

  /**
   * @internal
   */
  get name() {
    return 'playback_rate'
  }

  /**
   * @internal
   */
  get supportedVersion() {
    return { min: CLAPPR_VERSION }
  }

  private static readonly buttonTemplate = template(buttonHtml)

  private static readonly listTemplate = template(listHtml)

  constructor(core: Core) {
    super(core)
    this.playbackRates =
      core.options.playbackRate?.options || DEFAULT_PLAYBACK_RATES
    this.selectedRate =
      core.options.playbackRate?.defaultValue || DEFAULT_PLAYBACK_RATE
  }

  /**
   * @internal
   */
  override get attributes() {
    return {
      class: 'media-control-playbackrate',
    }
  }

  /**
   * @internal
   */
  override get events() {
    return {
      'click .gear-sub-menu_btn': 'onRateSelect',
      'click .gear-option': 'onShowMenu',
      'click .go-back': 'goBack',
    }
  }

  /**
   * @internal
   */
  override bindEvents() {
    this.listenTo(this.core, Events.CORE_READY, this.onCoreReady)
    this.listenTo(
      this.core,
      Events.CORE_ACTIVE_CONTAINER_CHANGED,
      this.onActiveContainerChange,
    )
  }

  private onCoreReady() {
    trace(`${T} onCoreReady`)
    const mediaControl = this.core.getPlugin('media_control')
    assert(mediaControl, 'media_control plugin is required')
    const gear = this.core.getPlugin('bottom_gear') as BottomGear
    assert(gear, 'bottom_gear plugin is required')
    this.listenTo(
      mediaControl,
      MediaControlEvents.MEDIACONTROL_GEAR_RENDERED,
      this.onGearRendered,
    )
  }

  private onActiveContainerChange() {
    trace(`${T} onActiveContainerChange`)
    this.listenTo(this.core.activePlayback, Events.PLAYBACK_STOP, this.onStop)
    this.listenTo(this.core.activePlayback, Events.PLAYBACK_PLAY, this.onPlay)
    this.listenTo(
      this.core.activePlayback,
      PlaybackEvents.PLAYBACK_RATE_CHANGED,
      this.onPlaybackRateChange,
    )
    this.listenTo(
      this.core.activeContainer,
      Events.CONTAINER_PLAYBACKDVRSTATECHANGED,
      this.onDvrStateChanged,
    )
  }

  private onGearRendered() {
    trace(`${T} onGearRendered`, {
      rendered: this.rendered,
    })
    this.rendered = false
    this.render()
  }

  private onDvrStateChanged(dvrEnabled: boolean) {
    trace(`${T} onDvrStateChanged`, {
      dvrEnabled,
    })
    if (dvrEnabled) {
      this.render()
    }
  }

  private allRateElements(): ZeptoResult {
    return this.$('ul.gear-sub-menu li')
  }

  private rateElement(rate = '1'): ZeptoResult {
    return (
      this.$(`ul.gear-sub-menu a[data-rate="${rate}"]`) as ZeptoResult
    ).parent()
  }

  private onPlaybackRateChange(playbackRate: number) {
    const selectedRate = parseInt(this.selectedRate, 10)
    // TODO check it doesn't interfere with the DASH.js or HLS.js playback live catchup
    if (Math.abs(playbackRate - selectedRate) > 0.1) {
      trace(`${T} onPlaybackRateChange setting target rate`, {
        playbackRate,
        selectedRate,
      })
      this.core.activePlayback?.setPlaybackRate(selectedRate)
    }
  }

  private shouldRender() {
    if (!this.core.activePlayback) {
      return false
    }

    if (
      this.core.activePlayback.getPlaybackType() === Playback.LIVE &&
      !this.core.activePlayback.dvrEnabled
    ) {
      return false
    }

    return 'setPlaybackRate' in this.core.activePlayback
  }

  /**
   * @internal
   */
  override render() {
    trace(`${T} render`, {
      rendered: this.rendered,
      shouldRender: this.shouldRender(),
    })

    if (!this.shouldRender()) {
      return this
    }

    if (this.rendered) {
      return this
    }

    const button = PlaybackRate.buttonTemplate({
      title: this.getTitle(),
      speedIcon,
      arrowRightIcon,
      i18n: this.core.i18n,
    })

    this.$el.html(button)

    ;(this.core.getPlugin('bottom_gear') as BottomGear)
      ?.getElement('rate')
      ?.html(this.el)

    this.rendered = true

    return this
  }

  private onStartAd() {
    this.prevSelectedRate = this.selectedRate
    this.resetPlaybackRate()
    this.listenToOnce(
      this.core.activePlayback,
      Events.PLAYBACK_PLAY,
      this.onFinishAd,
    )
  }

  private onFinishAd() {
    if (this.prevSelectedRate) {
      this.setSelectedRate(this.prevSelectedRate)
    }
  }

  private onPlay() {
    if (
      this.core.getPlaybackType() === Playback.LIVE &&
      !this.core.activePlayback.dvrEnabled
    ) {
      this.resetPlaybackRate()
    } else {
      this.setSelectedRate(this.selectedRate)
    }
  }

  private resetPlaybackRate() {
    this.setSelectedRate(DEFAULT_PLAYBACK_RATE)
  }

  private onStop() {}

  private onRateSelect(event: MouseEvent) {
    event.stopPropagation()
    const rate = (event.currentTarget as HTMLElement).dataset.rate
    if (rate) {
      this.setSelectedRate(rate)
      this.highlightCurrentRate()
    }

    return false
  }

  private onShowMenu() {
    this.$el.html(
      PlaybackRate.listTemplate({
        playbackRates: this.playbackRates,
        arrowLeftIcon,
        checkIcon,
        i18n: this.core.i18n,
      }),
    )
    ;(this.core.getPlugin('bottom_gear') as BottomGear)?.setContent(this.el)
    this.highlightCurrentRate()
  }

  private goBack() {
    setTimeout(() => {
      this.core.getPlugin('bottom_gear').refresh()
    }, 0)
  }

  private setSelectedRate(rate: string) {
    // Set <video playbackRate="..."
    this.core.activePlayback?.setPlaybackRate(rate)
    this.selectedRate = rate
  }

  private getTitle() {
    return (
      this.playbackRates.find((r) => r.value === this.selectedRate)?.label ||
      this.selectedRate
    )
  }

  private highlightCurrentRate() {
    this.allRateElements().removeClass('current')
    this.allRateElements().find('a').removeClass('gcore-skin-active')

    const currentLevelElement = this.rateElement(this.selectedRate)

    currentLevelElement.addClass('current')
    currentLevelElement.find('a').addClass('gcore-skin-active')
  }
}
