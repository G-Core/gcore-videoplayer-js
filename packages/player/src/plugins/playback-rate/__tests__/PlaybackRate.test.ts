import { describe, it, expect, beforeEach, vi, afterEach } from 'vitest'
import { PlaybackRate } from '../PlaybackRate'
import {
  createMockBottomGear,
  createMockCore,
  createMockMediaControl,
} from '../../../testUtils'
import { Events } from '@clappr/core'
import { GearEvents } from '../../bottom-gear/BottomGear'
// import { Logger, LogTracer, setTracer } from '@gcorevideo/utils'

// Logger.enable('*')
// setTracer(new LogTracer('PlaybackRate.test'))

describe('PlaybackRate', () => {
  let core: any
  let bottomGear: any
  let mediaControl: any
  let playbackRate: PlaybackRate
  beforeEach(() => {
    core = createMockCore()
    mediaControl = createMockMediaControl(core)
    bottomGear = createMockBottomGear(core)
    core.getPlugin.mockImplementation((name: string) => {
      if (name === 'bottom_gear') {
        return bottomGear
      }
      if (name === 'media_control') {
        return mediaControl
      }
      return null
    })
  })
  describe('basically', () => {
    beforeEach(() => {
      playbackRate = new PlaybackRate(core)
      core.emit(Events.CORE_READY)
      mediaControl.trigger(Events.MEDIACONTROL_RENDERED)
      core.activePlayback.getPlaybackType.mockReturnValue('live')
      core.activeContainer.getPlaybackType.mockReturnValue('live')
      core.getPlaybackType.mockReturnValue('live')
      core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED)
      bottomGear.trigger(GearEvents.RENDERED)
    })
    it('should render', () => {
      expect(playbackRate.el.innerHTML).toMatchSnapshot()
    })
    it('should have normal rate initially', () => {
      expect(
        playbackRate.$el.find('[data-rate="1"]').parent().hasClass('current'),
      ).toBe(true)
      expect(
        playbackRate.$el.find('[data-rate="1"]').hasClass('gcore-skin-active'),
      ).toBe(true)
    })
    describe('until media source is loaded', () => {
      it('should not attach to the media control', () => {
        expect(bottomGear.addItem).not.toHaveBeenCalledWith('rate', expect.anything())
      })
    })
    describe('after media source is loaded', () => {
      describe('when DVR is available', () => {
        beforeEach(() => {
          core.activePlayback.dvrEnabled = true
          core.activeContainer.isDvrEnabled.mockReturnValue(true)
          core.activePlayback.emit(Events.PLAYBACK_LOADEDMETADATA)
          core.activeContainer.emit(Events.CONTAINER_LOADEDMETADATA)
        })
        it('should attach to the media control', () => {
          expect(bottomGear.addItem).toHaveBeenCalledWith('rate', playbackRate.$el)
          expect(
            bottomGear.$el.find('li[data-rate]').text(),
            // @ts-ignore
          ).toMatchPlaybackRateLabel('1x')
        })
      })
      describe('when DVR is not available', () => {
        beforeEach(() => {
          core.activePlayback.emit(Events.PLAYBACK_LOADEDMETADATA)
          core.activeContainer.emit(Events.CONTAINER_LOADEDMETADATA)
        })
        it('should not attach to the media control', () => {
          expect(bottomGear.addItem).not.toHaveBeenCalledWith('rate', expect.anything())
        })
      })
    });
    describe('on playback rate select', () => {
      beforeEach(() => {
        core.activePlayback.dvrEnabled = true
        core.activeContainer.isDvrEnabled.mockReturnValue(true)
        core.activePlayback.emit(Events.PLAYBACK_LOADEDMETADATA)
        core.activeContainer.emit(Events.CONTAINER_LOADEDMETADATA)
      })
      describe.each([[2], [1.5], [1.25], [0.75], [0.5]])('%s', (rate) => {
        beforeEach(() => {
          playbackRate.$el.find(`[data-rate="${rate}"]`).click()
        })
        it('should set the selected rate', () => {
          expect(core.activePlayback.setPlaybackRate).toHaveBeenCalledWith(rate)
        })
        it('should highlight the selected rate', () => {
          expect(
            playbackRate.$el
              .find(`[data-rate="${rate}"]`)
              .parent()
              .hasClass('current'),
          ).toBe(true)
          expect(
            playbackRate.$el
              .find(`[data-rate="${rate}"]`)
              .hasClass('gcore-skin-active'),
          ).toBe(true)
        })
        it('should update the gear box option label', () => {
          expect(
            bottomGear.$el.find('#playback-rate-button').text(),
            // @ts-ignore
          ).toMatchPlaybackRateLabel(`${rate}x`)
        })
      })
    })
    describe('on go back', () => {
      beforeEach(async () => {
        playbackRate.$el.find('#playback-rate-back-button').click()
        return new Promise((resolve) => setTimeout(resolve, 0))
      })
      it('should refresh the bottom gear', () => {
        expect(bottomGear.refresh).toHaveBeenCalled()
      })
    })
  })
  describe('options.defaultValue', () => {
    beforeEach(() => {
      core.options.playbackRate = {
        defaultValue: 1.5,
      }
      playbackRate = new PlaybackRate(core)
      core.emit(Events.CORE_READY)
      mediaControl.trigger(Events.MEDIACONTROL_RENDERED)
      core.activePlayback.getPlaybackType.mockReturnValue('live')
      core.activeContainer.getPlaybackType.mockReturnValue('live')
      core.getPlaybackType.mockReturnValue('live')
      core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED)
      bottomGear.trigger(GearEvents.RENDERED)
      core.activePlayback.dvrEnabled = true
      core.activeContainer.isDvrEnabled.mockReturnValue(true)
      core.activePlayback.emit(Events.PLAYBACK_LOADEDMETADATA)
      core.activeContainer.emit(Events.CONTAINER_LOADEDMETADATA)
    })
    it('should set the selected rate to the defaultValue', () => {
      expect(core.activePlayback.setPlaybackRate).toHaveBeenCalledWith(1.5)
    })
    it('should highlight the selected rate', () => {
      expect(
        playbackRate.$el.find('[data-rate="1.5"]').parent().hasClass('current'),
      ).toBe(true)
      expect(
        playbackRate.$el.find('[data-rate="1.5"]').hasClass('gcore-skin-active'),
      ).toBe(true)
    })
    it('should render proper gear box option label', () => {
      expect(
        bottomGear.$el.find('#playback-rate-button').text(),
        // @ts-ignore
      ).toMatchPlaybackRateLabel('1.5x')
    })
  })
})

expect.extend({
  toMatchPlaybackRateLabel(received, expected) {
    const { isNot } = this
    return {
      pass:
        received
          .replace(/\/assets.*\.svg/g, '')
          .replace(/\s+/g, ' ')
          .trim().includes(`playback_rate ${expected}`),
      message: () => `${received} does${isNot ? '' : ' not'} match ${expected}`,
    }
  },
})
