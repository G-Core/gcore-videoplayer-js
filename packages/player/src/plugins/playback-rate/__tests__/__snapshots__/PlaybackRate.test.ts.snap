// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlaybackRate > basically > should render 1`] = `
"<button class="gplayer-lite-btn go-back gcore-skin-text-color" id="playback-rate-back-button">
    <span class="arrow-left-icon">/assets/icons/new/arrow-left.svg</span>
    playback_rate
</button>
<ul class="gear-sub-menu" id="playback-rate-menu">
    
        <li>
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-rate="0.5">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                0.5x
            </a>
        </li>
    
        <li>
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-rate="0.75">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                0.75x
            </a>
        </li>
    
        <li class="current">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color gcore-skin-active" data-rate="1">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                1x
            </a>
        </li>
    
        <li>
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-rate="1.25">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                1.25x
            </a>
        </li>
    
        <li>
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-rate="1.5">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                1.5x
            </a>
        </li>
    
        <li>
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-rate="1.75">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                1.75x
            </a>
        </li>
    
        <li>
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-rate="2">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                2x
            </a>
        </li>
    
</ul>"
`;
